### PR TITLE
keep contrast using droplevels

### LINKED ISSUE #454: When `AVISIT` is an ordered factor and I manually set contrasts, my custom contrasts seem to be ignored:
When `AVISIT` is an ordered factor and I manually set contrasts, my custom contrasts seem to be ignored:

``` r
library(mmrm)
packageVersion("mmrm")
#> [1] '0.3.12.9000'
packageDescription("mmrm")$GithubSHA1
#> [1] "b06da426f04eb6bb4edbc17a9159c89020030fca"
data("fev_data", package = "mmrm")
fev_data$AVISIT <- ordered(fev_data$AVISIT)
contrasts(fev_data$AVISIT) <- contr.treatment(n = 4)
coef(lm(FEV1 ~ AVISIT, data = fev_data))
#> (Intercept)     AVISIT2     AVISIT3     AVISIT4 
#>   34.605646    5.159606   10.237350   15.572446
coef(mmrm(FEV1 ~ AVISIT + us(AVISIT | USUBJID), data = fev_data))
#> (Intercept)    AVISIT.L    AVISIT.Q    AVISIT.C 
#>  42.3854364  11.5452671   0.5005517   0.1479297
```

<sup>Created on 2024-07-09 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value
#>  version  R version 4.4.0 (2024-04-24)
#>  os       macOS Sonoma 14.5
#>  system   aarch64, darwin20
#>  ui       X11
#>  language (EN)
#>  collate  en_US.UTF-8
#>  ctype    en_US.UTF-8
#>  tz       America/Indiana/Indianapolis
#>  date     2024-07-09
#>  pandoc   3.1.11 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/aarch64/ (via rmarkdown)
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  package     * version     date (UTC) lib source
#>  backports     1.4.1       2021-12-13 [1] CRAN (R 4.4.0)
#>  checkmate     2.3.1       2023-12-04 [1] CRAN (R 4.4.0)
#>  cli           3.6.2       2023-12-11 [1] CRAN (R 4.4.0)
#>  digest        0.6.35      2024-03-11 [1] CRAN (R 4.4.0)
#>  evaluate      0.23        2023-11-01 [1] CRAN (R 4.4.0)
#>  fansi         1.0.6       2023-12-08 [1] CRAN (R 4.4.0)
#>  fastmap       1.2.0       2024-05-15 [1] CRAN (R 4.4.0)
#>  fs            1.6.4       2024-04-25 [1] CRAN (R 4.4.0)
#>  generics      0.1.3       2022-07-05 [1] CRAN (R 4.4.0)
#>  glue          1.7.0       2024-01-09 [1] CRAN (R 4.4.0)
#>  htmltools     0.5.8.1     2024-04-04 [1] CRAN (R 4.4.0)
#>  knitr         1.46        2024-04-06 [1] CRAN (R 4.4.0)
#>  lattice       0.22-6      2024-03-20 [1] CRAN (R 4.4.0)
#>  lifecycle     1.0.4       2023-11-07 [1] CRAN (R 4.4.0)
#>  magrittr      2.0.3       2022-03-30 [1] CRAN (R 4.4.0)
#>  Matrix        1.7-0       2024-03-22 [1] CRAN (R 4.4.0)
#>  mmrm        * 0.3.12.9000 2024-07-09 [1] Github (openpharma/mmrm@b06da42)
#>  nlme          3.1-164     2023-11-27 [1] CRAN (R 4.4.0)
#>  pillar        1.9.0       2023-03-22 [1] CRAN (R 4.4.0)
#>  pkgconfig     2.0.3       2019-09-22 [1] CRAN (R 4.4.0)
#>  purrr         1.0.2       2023-08-10 [1] CRAN (R 4.4.0)
#>  R.cache       0.16.0      2022-07-21 [1] CRAN (R 4.4.0)
#>  R.methodsS3   1.8.2       2022-06-13 [1] CRAN (R 4.4.0)
#>  R.oo          1.26.0      2024-01-24 [1] CRAN (R 4.4.0)
#>  R.utils       2.12.3      2023-11-18 [1] CRAN (R 4.4.0)
#>  rbibutils     2.2.16      2023-10-25 [1] CRAN (R 4.4.0)
#>  Rcpp          1.0.12      2024-01-09 [1] CRAN (R 4.4.0)
#>  Rdpack        2.6         2023-11-08 [1] CRAN (R 4.4.0)
#>  reprex        2.1.0       2024-01-11 [1] CRAN (R 4.4.0)
#>  rlang         1.1.4       2024-06-04 [1] CRAN (R 4.4.0)
#>  rmarkdown     2.26        2024-03-05 [1] CRAN (R 4.4.0)
#>  rstudioapi    0.16.0      2024-03-24 [1] CRAN (R 4.4.0)
#>  sessioninfo   1.2.2       2021-12-06 [1] CRAN (R 4.4.0)
#>  stringi       1.8.4       2024-05-06 [1] CRAN (R 4.4.0)
#>  stringr       1.5.1       2023-11-14 [1] CRAN (R 4.4.0)
#>  styler        1.10.3      2024-04-07 [1] CRAN (R 4.4.0)
#>  tibble        3.2.1       2023-03-20 [1] CRAN (R 4.4.0)
#>  TMB           1.9.14      2024-07-03 [1] CRAN (R 4.4.0)
#>  utf8          1.2.4       2023-10-22 [1] CRAN (R 4.4.0)
#>  vctrs         0.6.5       2023-12-01 [1] CRAN (R 4.4.0)
#>  withr         3.0.0       2024-01-16 [1] CRAN (R 4.4.0)
#>  xfun          0.43        2024-03-25 [1] CRAN (R 4.4.0)
#>  yaml          2.3.8       2023-12-11 [1] CRAN (R 4.4.0)
#> 
#>  [1] /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library
#> 
#> ──────────────────────────────────────────────────────────────────────────────
```

</details>

I would like to handle time as an ordered factor because the ordering of the levels is important for ARMA correlation structures. Unfortunately, [R uses `contr.poly()` on ordered factors by default](https://library.virginia.edu/data/articles/understanding-ordered-factors-in-a-linear-model), which is usually counterintuitive and often statistically inappropriate. Would it be reasonable for `mmrm` to respect user-defined contrasts so users like me can work around the strange defaults in base R?

_Originally posted by @wlandau in https://github.com/openpharma/mmrm/discussions/451_

### PR DESCRIPTION
**Pull request by clarkliming**
close #454 

the contrasts can be preserved now. (but how will it work if new data does not contain the same contrast while doing predictions?)

### CODE REVIEW

**Review by danielinteractive — COMMENTED** [review]
Thanks @clarkliming , interesting - can you as a next step add a test that shows the preserved contrasts feature? Also please add a `NEWS` entry.
Just wondering if we really need to replace the base R function? Is this a common issue?

**Comment by clarkliming** [comment]
previously I was using another approach to ensure the factor levels are consistent. Now i have another solution, and `h_factor_ref` can be removed now. @danielinteractive please have a look

File: R/tmb.R [hunk 1]
```diff
@@ -145,11 +146,28 @@ h_mmrm_tmb_data <- function(formula_parts,
...
       data = data,
       weights = .(as.symbol(weights_name)),
-      na.action = "na.pass"
+      na.action = "na.pass",
+      xlev = xlev
     ))
   )
   if (drop_levels) {
-    full_frame <- droplevels(full_frame, except = formula_parts$visit_var)
+    ix <- vapply(
```
Comment (danielinteractive) [inline R/tmb.R:154]: this part here gets quite complex, can you factor this out as a helper function with unit tests?

**Review by danielinteractive — COMMENTED** [review]
Thanks a lot @clarkliming ! That looks much better, good that it gets simpler overall now. Just one comment below

File: R/tmb.R (hunk 1 above)
Comment (clarkliming) [reply R/tmb.R]: now it is using `droplevels` again, but added a helper to check if a factor should drop levels

**Comment by clarkliming** [comment]
after checking the current implementation, I kind want to expose this "droplevels" option to users, so that if they really want to automatically drop levels to avoid singular matrix it is possible; but if they want to use some specific contrasts to construct the model matrix (even though they might still be singular) it is still possible. (otherwise if calling droplevels the contrasts are totally lost)

**Comment by danielinteractive** [comment]
> after checking the current implementation, I kind want to expose this "droplevels" option to users, so that if they really want to automatically drop levels to avoid singular matrix it is possible; but if they want to use some specific contrasts to construct the model matrix (even though they might still be singular) it is still possible. (otherwise if calling droplevels the contrasts are totally lost)

How is this handled in other model fitting functions in R? Just thinking it would be nice to keep with the tradition, if there is one.

**Comment by clarkliming** [comment]
> > after checking the current implementation, I kind want to expose this "droplevels" option to users, so that if they really want to automatically drop levels to avoid singular matrix it is possible; but if they want to use some specific contrasts to construct the model matrix (even though they might still be singular) it is still possible. (otherwise if calling droplevels the contrasts are totally lost)
> 
> How is this handled in other model fitting functions in R? Just thinking it would be nice to keep with the tradition, if there is one.

in `stats`, `lm` and `glm` will use `model.frame(drop.unused.levels = T)` to drop these levels, including contrasts. However we are doing differently because the levels for `visit_var` should be kept/discarded at users' wish. What we have now is already very similar to what `stats` have.  exposing this `droplevels` will just add another flexibility but maybe too minor.

**Review by danielinteractive — COMMENTED** [review]
Thanks @clarkliming , yeah I think we can live without the additional user argument for now

File: R/tmb.R [hunk 2]
```diff
@@ -145,11 +147,29 @@ h_mmrm_tmb_data <- function(formula_parts,
...
       data = data,
       weights = .(as.symbol(weights_name)),
-      na.action = "na.pass"
+      na.action = "na.pass",
+      xlev = xlev
     ))
   )
   if (drop_levels) {
-    full_frame <- droplevels(full_frame, except = formula_parts$visit_var)
+    all_cols <- colnames(full_frame)
```
[the remark this answers is not in the document]
Comment (clarkliming) [reply R/tmb.R]: this part is updated; although this part is not more complicated than other parts in this function (do we need to also factor other part of this function?)

File: tests/testthat/test-fit.R [hunk 3]
```diff
@@ -470,6 +479,19 @@ test_that("divergent optimizer will not be marked as success", {
   )
 })
 
+test_that("mmrm will use the provided contrast", {
+  fev2 <- fev_data
+  contrasts(fev2$ARMCD) <- contr.poly(n = 2)
+  fit <- mmrm(FEV1 ~ ARMCD * AVISIT + ar1(AVISIT | USUBJID), data = fev2)
+  expect_identical(
+    names(coef(fit)),
```
[the remark this answers is not in the document]
Comment (clarkliming) [reply tests/testthat/test-fit.R]: snapshot added. the value itself is not checked (it seems SAS do not provide such flexibility).

File: R/tmb-methods.R [hunk 4]
```diff
@@ -79,16 +79,16 @@ predict.mmrm_tmb <- function(object,
...
     object$formula_parts, newdata,
     weights = rep(1, nrow(newdata)),
     reml = TRUE,
     singular = "keep",
     drop_visit_levels = FALSE,
     allow_na_response = TRUE,
-    drop_levels = FALSE
+    drop_levels = FALSE,
+    xlev = stats::.getXlevels(terms(object), object$tmb_data$full_frame),
+    contrasts = attr(object$tmb_data$x_matrix, "contrasts")
```
[the remark this answers is not in the document]
Comment (clarkliming) [reply R/tmb-methods.R]: added in component

File: R/tmb-methods.R [hunk 5]
```diff
@@ -79,16 +79,16 @@ predict.mmrm_tmb <- function(object,
...
   tmb_data <- h_mmrm_tmb_data(
     object$formula_parts, newdata,
     weights = rep(1, nrow(newdata)),
     reml = TRUE,
     singular = "keep",
     drop_visit_levels = FALSE,
     allow_na_response = TRUE,
-    drop_levels = FALSE
+    drop_levels = FALSE,
+    xlev = stats::.getXlevels(terms(object), object$tmb_data$full_frame),
```
[the remark this answers is not in the document]
Comment (clarkliming) [reply R/tmb-methods.R]: added in component

**Comment by clarkliming** [comment]
may I ask for what reason we are calling `model.frame` multiple times here? https://github.com/openpharma/mmrm/blob/424810d88b0507e9e88242b61d343852358b2df2/R/tmb-methods.R#L233

**Comment by danielinteractive** [comment]
> may I ask for what reason we are calling `model.frame` multiple times here?
> 
> https://github.com/openpharma/mmrm/blob/424810d88b0507e9e88242b61d343852358b2df2/R/tmb-methods.R#L233

Good question, it took me 1 minute to see it, but in line https://github.com/openpharma/mmrm/blob/424810d88b0507e9e88242b61d343852358b2df2/R/tmb-methods.R#L237
the `formula_full` is used instead of `formula` further up. So it is a slightly different call. 

If you have an idea how to simplify this method that would be great of course, I checked and see that we have good unit testing coverage of it, which will help with refactoring.

**Comment by clarkliming** [comment]
> > may I ask for what reason we are calling `model.frame` multiple times here?
> > https://github.com/openpharma/mmrm/blob/424810d88b0507e9e88242b61d343852358b2df2/R/tmb-methods.R#L233
> 
> Good question, it took me 1 minute to see it, but in line
> 
> https://github.com/openpharma/mmrm/blob/424810d88b0507e9e88242b61d343852358b2df2/R/tmb-methods.R#L237
> 
> 
> the `formula_full` is used instead of `formula` further up. So it is a slightly different call.
> If you have an idea how to simplify this method that would be great of course, I checked and see that we have good unit testing coverage of it, which will help with refactoring.

probably we can use the full frame, and then subset it to the column we have in the non-full frame? that will keep the subjects available in full frame.  but anyway this is not that important, other than calling model.frame will return multiple "contrasts are lost thing" (this pops up when new data is used; but we treated all data as new even if we use the data in the mmrm fit)

**Comment by danielinteractive** [comment]
OK, understood, yeah maybe in a second PR could be simplified to reduce number of warnings.

**Review by danielinteractive — APPROVED** [review]
looks great, thanks @clarkliming

**Pull request merged**